tests|github|pre-commit: Improve `build-workflows.py` and make it a `pre-commit` hook

xdustinface opened the pull request:
This PR:

- Implements some general improvements in `build-workflows.py`.
- Disables some tests which currently don't work.
- Forces `build-workflows.py` to run on CI and throw an error if there are is a workflow upgrade required.
- Changes `build-workflows.py` so that we don't need manually add new test directories to `testconfig.py` but instead it just uses all directories which contain a `test_*.py` file.
- Simplifies the generated workflows so that we don't need to explicitly list all test files.

For details see individual commit messages.

xdustinface committed:
tests: Move `default_replacements` into `generate_replacements`

xdustinface committed:
tests: Directly use `root_test_dirs`

Instead of passing it into `subdirs`

xdustinface committed:
tests: Make `build-workflow.py` callable from everywhere

xdustinface committed:
tests: Adjust the output of `build-workflows.py`

xdustinface committed:
tests: Add `-f` to `build-workflows.py` to allow failing on CI

xdustinface committed:
pre-commit: Run `build-workflows.py` as first hook

xdustinface committed:
github: Run `build-workflows.py`

arvidn commented on .github/workflows/build-test-macos-blockchain.yml:
```diff
@@ -28,7 +28,7 @@ jobs:
     steps:
     - name: Cancel previous runs on the same branch
       if: ${{ github.ref != 'refs/heads/main' }}
-      uses: styfle/cancel-workflow-action@0.9.1
```
I think you should update the template with version `0.9.1`, then all of these won't regress

xdustinface replied:
Good point, changed! I didn't even realized that this are downgrades :see_no_evil:

arvidn commented on tests/build-workflows.py:
```diff
@@ -9,12 +9,18 @@
+def subdirs() -> List[Path]:
+    result: List[Path] = []
+    for r in testconfig.root_test_dirs:
+        for d in Path(root_path / r).rglob("**/"):
```
it seems much safer if we rely on `pytest` to do the globbing, rather than this file having to list all the tests. richard brought this up, that we should just specify the `tests` directory, or perhaps `tests/test_*.py`

xdustinface replied:
Yeah that's probably better but rather out of scope of this PR.

arvidn commented on .pre-commit-config.yaml:
```diff
@@ -1,4 +1,11 @@
+    hooks:
+    -   id: build-workflows
+        name: Validate github action workflows
+        entry: ./tests/build-workflows.py -f True
```
does this run in CI as well, to prevent PRs that fail to run `./build-workflows.py` when they would need to?

xdustinface replied:
Yes see https://github.com/Chia-Network/chia-blockchain/pull/8728/checks?check_run_id=3805955316#step:4:70

aqk reviewed (dismissed):
This is great. Only issue I see is that this patch introduces globals into several functions.

xdustinface commented:
Turned this into draft for now. Want to get #8811 merged first and it seems like it still takes some more time until all tests will pass so that we don't need a feature to exclude tests.

xdustinface committed:
tests: Skip DID wallet tests until they are reliable and fixed

xdustinface committed:
tests: Skip `test_using_legacy_cryptfilekeyring`

xdustinface committed:
tests|github: Drop `root_test_dirs` and just search in all directories

This removes the requirement to add new test directories to the 
`testconfig.py`. This change uncovered two more tests which were not 
enabled on CI because their parent dir's were not added to 
`root_test_dirs`.
- `tests/weight_proof/test_weight_proof.py`
- `tests/util/test_struct_stream.py`

xdustinface committed:
test: Improve workflow update checks by comparing old vs new

Instead of doing a `git diff` which can lead to false positives if there
are other local changes unlrelated to test workflows. This also fixes
the issue that newly added files were not tracked by `git diff` which
means the script / the CI would trigger an alert.

xdustinface committed:
tests|github: Don't list test files, use `tests/dir/tests_*.py` instead

xdustinface committed:
tests: Adjust `build-workflows.py` to work as expected on windows

xdustinface committed:
tests: Read/Write in bytes to avoid and detect unexpected `\r\n`

altendky reviewed (changes requested):
Alrighty, a few more minor tweaks (all in suggestions I think) and now I think I've looked over everything mostly.

aqk reviewed (dismissed):
Looks great!

richardkiss reviewed (dismissed):
I'm still not a fan of proliferation of workflow files vs `use-matrix`, but this change is unquestionably an improvement on the state of things.

xdustinface committed:
tests: Skip `test_pool_cmdline.py`

altendky reviewed (approved):
I agree with Adam's comments, and could nitpick some stuff myself...  but mostly I agree with Arvid that this needs to get in.  It looks like in the past couple weeks we have a few more tests that haven't been being run and are failing.
